Insert into DeletedCouchDoc if change is deletion

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-14584

Now that the `DeletedCouchDoc` SQL table exists, we need to make use of it. Fortunately, there is one place in code where all couch changes pass through and that is the `KafkaProcessor` in pillows, responsible for publishing couch changes to kafka.

I thought about creating another processor dedicated to this and setting up multiple processors on the couch -> kafka pillows, but that felt like overkill for what is a fairly simple change.

Instead I added a few lines of code to the existing processor that use the change metadata `original_publication_datetime` field as the `deleted_on` field for the `DeletedCouchDoc` entry.

#### Concerns
I know we have undo workflows, but I don't think those apply to couch docs. I need to investigate that a bit more.
- [x] investigate how this behaves when undoing deletions (if that applies to couch docs)

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
This is a fairly straightforward addition to the KafkaProcessor object to write deletions to the DeletedCouchDoc. Following guidelines to keep pillows idempotent, if an object with the same `doc_id` and `doc_type` already exists in `DeletedCouchDoc`, then swallow the error and move on.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
insert into DeletedCouchDoc if change is deletion

biyeun commented on corehq/apps/change_feed/pillow.py:
```diff
@@ -98,3 +106,13 @@ def _get_doc_type_from_change(change):
+                                       doc_type=doc_type,
+                                       deleted_on=deleted_on)
+    except IntegrityError:
+        # if it already exists, ignore it
```
can you think of any occasion where a change would be picked up to `is_deletion` on an already deleted doc? would we want to know if this keeps happening?

gherceg replied:
I was actually just talking to Cal about how pillows are designed to be idempotent, so the straightforward example for your question is this pillow being rewound to an earlier checkpoint and processing this change again. That was the main reason I chose to just ignore it entirely and not flag anything, since by design pillows should be able to handle this.

biyeun replied:
got it 👍

millerdev committed:
Fix bug in KafkaProcessor.process_change

millerdev committed:
Merge branch 'master' into gh/data-deletion/soft-delete-couch-docs

millerdev committed:
Handle null doc_type in KafkaProcessor

Fixes error in AppPillowTest.test_hard_delete_app

TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.

which was caused by

IntegrityError: null value in column "doc_type" of relation "cleanup_deletedcouchdoc" violates not-null constraint

millerdev committed:
Remove DeletedCouchDoc record on undo delete

millerdev committed:
Lint

millerdev committed:
Remove DeletedCouchDoc on undo delete Application

millerdev committed:
Auto-cleanup DeleteRecord docs

millerdev commented:
Couch models like `Application` that have undo deletion workflows should be reviewed to determine if it is necessary to add more commits similar to https://github.com/dimagi/commcare-hq/pull/33040/commits/2e283989880473a59c01454b30d756e43fa26325.

millerdev committed:
Merge branch 'master' into gh/data-deletion/soft-delete-couch-docs

minhaminha committed:
Add undo handlers to delete associated DeletedCouchDoc for forms, groups, web users and modules + add tests to cover all those undo processes + fix slight error in app related undo process

millerdev commented on corehq/apps/app_manager/tests/test_apps.py:
```diff
@@ -127,7 +127,7 @@ def test_undo_delete_app_removes_deleted_couch_doc_record(self):
         rec.undo()
         with self.assertRaises(DeletedCouchDoc.DoesNotExist):
-            DeletedCouchDoc.objects.get(doc_id=self.app._id, doc_type=self.app.doc_type)
+            DeletedCouchDoc.objects.get(doc_id=rec._id, doc_type=rec.doc_type)
```
I think `rec.doc_type` is incorrect here since it will have switched back to the non-deleted version after `rec.undo()`, however the test will still pass because a `DeletedCouchDoc` record will not exist for `doc_type="Application"`. I think this should use a literal like `"Application-Deleted"` to avoid accidentally using the wrong query parameter. Alternately, use the slightly safer pattern in the next suggestion.

minhaminha replied:
`rec.doc_type` here refers to `DeleteApplicationRecord` not the actual application that's been deleted. Definitely confused me too since both Applications and Groups append '-Deleted' to the deleted objects' doc_type and feels like that should be used but like with all the other models that inherit `DeleteRecord` its the Delete(Model)Record's _id and doc_type that's used to create the DeleteCouchDoc

millerdev commented on corehq/apps/app_manager/tests/test_apps.py:
```diff
@@ -136,10 +136,26 @@ def testDeleteForm(self):
+    def test_undo_delete_form_removes_deleted_couch_doc_record(self):
+        form = self.app.modules[0].forms[0]
+        rec = self.app.delete_form(self.app.modules[0].unique_id, form.unique_id)
+        assert DeletedCouchDoc.objects.get(doc_id=rec._id, doc_type=rec.doc_type)
+        rec.undo()
+        with self.assertRaises(DeletedCouchDoc.DoesNotExist):
+            DeletedCouchDoc.objects.get(doc_id=rec._id, doc_type=rec.doc_type)
```
Here is a safer pattern to avoid a hard-coded doc type literal, ensuring that the specific `DeletedCouchDoc` in question has been deleted regardless of what changes are made by `rec.undo()`.
```suggestion
        params = {"doc_id": rec._id, "doc_type": rec.doc_type}
        assert DeletedCouchDoc.objects.get(**params)
        rec.undo()
        with self.assertRaises(DeletedCouchDoc.DoesNotExist):
            DeletedCouchDoc.objects.get(**params)
```

minhaminha replied:
Made these changes to all the applicable tests!

minhaminha committed:
Modify previously written tests to be safer + fix a util test to pass

minhaminha committed:
Make rec an optional arugment for undo_delete

minhaminha committed:
Lint fixes

millerdev commented on corehq/ex-submodules/dimagi/utils/couch/undo.py:
```diff
@@ -91,11 +91,12 @@ def get_deleted_doc_type(document_class_or_instance):
-        doc_type=delete_record.doc_type,
-    ).delete()
+def undo_delete(document, delete_record=None, save=True):
+    if delete_record:
```
nit
```suggestion
    if delete_record is not None:
```

minhaminha committed:
apply suggested change

Co-authored-by: Daniel Miller <dmiller@dimagi.com>
